src/services/Batcher.ts

```diff
@@ -0,0 +1,102 @@
+import { MaybePromise } from '@/types'
+
+export type BatchCallbackLookup<V, R> = (value: V) => MaybePromise<R>
+export type BatchCallback<V, R> = (values: V[]) => MaybePromise<Map<V, R>> | BatchCallbackLookup<V, R>
```

Comment:
Just so I understand this properly; the `MaybePromise<Map<V, R>> | BatchCallbackLookup<V, R>` return value is meant to support a many to one AND a many to many model, right? And the many to many lookup is a `<MaybePromise>` just cause why not?

Reply:
So the map or the lookup can both return the correct value. One to many works either way as long as you set up the map or the lookup correctly. A map is more efficient but the lookup is more flexible. Being able to return an arrow function that knows how to get the value means we can support lots of different use cases.

Reply:
Makes sense to me, thanks!